File: java/leetcode_grind/Day858.java
package leetcode_grind;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Day858 {

    // https://leetcode.com/problems/minimum-operations-to-make-a-uni-value-grid/description/
    static class Solution1 {
        public int minOperations(int[][] grid, int x) {
            List<Integer> numsArray = new ArrayList<>();
            int result = 0;
            for (int row = 0; row < grid.length; row++) {
                for (int col = 0; col < grid[0].length; col++) {
                    numsArray.add(grid[row][col]);
                }
            }

            Collections.sort(numsArray);
            int length = numsArray.size();
            int finalCommonNumber = numsArray.get(length / 2);

            for (int number : numsArray) {
                if (number % x != finalCommonNumber % x) {
                    return -1;
                }
                result += Math.abs(finalCommonNumber - number) / x;
            }
            return result;
        }
    }

    static class Solution2 {
        public int minOperations(int[][] grid, int x) {
            List<Integer> numsArray = new ArrayList<>();
            int result = 0;

            for (int row = 0; row < grid.length; row++) {
                for (int col = 0; col < grid[0].length; col++) {
                    if (grid[row][col] % x != grid[0][0] % x)
                        return -1;
                    numsArray.add(grid[row][col]);
                }
            }

            Collections.sort(numsArray);
            int length = numsArray.size();
            int prefixIndex = 0;
            int suffixIndex = length - 1;

            while (prefixIndex < suffixIndex) {
                if (prefixIndex < length - suffixIndex - 1) {
                    int prefixOperations = ((prefixIndex + 1)
                            * (numsArray.get(prefixIndex + 1) - numsArray.get(prefixIndex))) / x;
                    result += prefixOperations;
                    prefixIndex++;
                } else {
                    int suffixOperations = ((length - suffixIndex)
                            * (numsArray.get(suffixIndex) - numsArray.get(suffixIndex - 1))) / x;
                    result += suffixOperations;
                    suffixIndex--;
                }
            }

            return result;
        }
    }
}

File: src/leetcode_grind/day_858.rs
// https://leetcode.com/problems/minimum-operations-to-make-a-uni-value-grid/description/
pub fn min_operations(grid: Vec<Vec<i32>>, x: i32) -> i32 {
    let mut nums_array = vec![];
    let mut result = i32::MAX;

    for row in 0..grid.len() {
        for col in 0..grid[0].len() {
            if grid[row][col] % x != grid[0][0] % x {
                return -1;
            }
            nums_array.push(grid[row][col]);
        }
    }

    nums_array.sort();
    let len = nums_array.len();
    let mut prefix_sum = vec![0; len];
    let mut suffix_sum = vec![0; len];

    for index in 1..len {
        prefix_sum[index] = prefix_sum[index-1] + nums_array[index - 1];
    } 

    for index in (0..len - 1).rev() {
        suffix_sum[index] = suffix_sum[index + 1] + nums_array[index + 1];
    }

    for index in 0..len {
        let left_operations = (nums_array[index] * index as i32 - prefix_sum[index]) / x;
        let right_operations = (suffix_sum[index] - nums_array[index] * (len - index - 1) as i32) / x;

        result = result.min(left_operations + right_operations);
    }

    result
}